### PR TITLE
Move to Dart Sass compiler

### LINKED ISSUE #22729: Scripts: Consider replacing `node-sass` with Dart SASS
## Description

Related to #21730 that added support for CSS imports in JavaScript files to `start` and `build` scripts in `@wordpress/scripts` package.

In https://github.com/WordPress/gutenberg/pull/21730#discussion_r427316779, @aduth raised the following issue related to `node-sass` package used:

> I hope we don't come to regret foisting this troublesome dependency upon consumers of @wordpress/scripts

Indeed, we had many reports when people struggled with this package when upgrading Node.js version because it requires compilation.

@yscik suggested the following:

> I would suggest changing this to Dart SASS (`sass` package). It's the [primary implementation](https://sass-lang.com/dart-sass) for SASS, and saves contributors all the native compilation trouble `node-sass` brings. `sass-loader` picks it up without any additional configuration.

If we decide to go this route, we should ensure we use the same approach in Gutenberg project consistently.

### PR DESCRIPTION
**Pull request by yscik**
Fixes #22729

## Description

* Replaces `node-sass` with `sass` ([Dart Sass](https://sass-lang.com/dart-sass)) as the SASS compiler implementation. — This is a drop-in replacement with the same API, but without having to compile native C bindings.

* Adds a patch for `react-dates`, renaming it's `_datepicker.css` file to the `scss` extension
  * This file is included by the DateTimePicker component
  * It has, for some reason, `darken()` function calls, which fails Dart Sass's CSS validation, as it's a Sass function
  * Changing it's file extension is the only way I found for it to be interpreted as Sass

## How has this been tested?

Comparing the output of the compiled CSS files, there are only stylistic differences:

* Whitespace and linebreak changes around block endings
* Double quotes around attribute values in attribute selectors
* Numeric precision (number of digits after decimals)  — "Dart Sass defaults to a sufficiently high precision for all existing browsers"

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] ~My code follows the accessibility standards. ~
- [ ] ~My code has proper inline documentation. ~
- [x] I've included developer documentation if appropriate. 
- [ ] ~I've updated all React Native files affected by any refactorings/renamings in this PR. ~

### CODE REVIEW

**Comment by gziolo** [comment]
@yscik, thank you for opening this PR. Let's make sure it gets some proper review. I will raise it as a topic in the next weekly WordPress Core JS chat.

**Comment by nerrad** [comment]
This was brought up in the weekly [WordPress Core JS chat](https://github.com/WordPress/gutenberg/pull/26450) (signup needed) and in general there's no opposition to this change. Will still need an approval from a review.

File: patches/react-dates+17.2.0.patch
Comment (gziolo) [inline patches/react-dates+17.2.0.patch:1]: I was wondering if there is a better way to fix the issue with `react-dates`. Would it be possible to whitelist somewhere in the webpack config instead?

Comment (yscik) [reply patches/react-dates+17.2.0.patch:1]: If I'm understanding the build process correctly, the packages are built outside of webpack with a [direct call to sass.render](https://github.com/WordPress/gutenberg/blob/e215c80e1fcb613dab3b26d9a31053ae96725226/bin/packages/build-worker.js#L99), so webpack's loader config is not used here. Haven't found a way to configure this in Dart-Sass, the syntax is [determined based on file extension.](https://github.com/sass/dart-sass/blob/82b2779d5caea545d1900b2a188d462cc68dfc03/lib/src/executable/compile_stylesheet.dart#L50-L56) 

Another approach that seems to work is creating a symlink in the `date-time` component:
` _datepicker.scss` ->  `../../../../node_modules/react-dates/lib/css/_datepicker.css`

Comment (gziolo) [reply patches/react-dates+17.2.0.patch:1]: https://github.com/WordPress/gutenberg/blob/87fa5f58534281aadd9bdf624f9b8384cfe66c89/packages/components/src/date-time/style.scss#L4

would it work if the extension was explicitly provided there?

Comment (yscik) [reply patches/react-dates+17.2.0.patch:1]: If it's `_datepicker.scss`, it won't find the file, if it's `_datepicker.css`, the issue remains.

Comment (gziolo) [reply patches/react-dates+17.2.0.patch:1]: This issue still exists in the latest version:

https://unpkg.com/browse/react-dates@21.8.0/lib/css/_datepicker.css

Maybe, we should simply copy this file to the component's folder and import the local version :)

**Comment by aristath** [comment]
node-sass is now [officially deprecated](https://sass-lang.com/blog/libsass-is-deprecated) so that's one more reason to go through with this 👍

**Review by gziolo — APPROVED** [review]
Let's move forward with this one and figure out what to do about `react-dates` CSS file separately.

**Comment by gziolo** [comment]
Opened #26534 with a proposal to copy SCSS file from `react-dates` to `@wordpress/components`.

**Comment by etoledom** [comment]
This PR seems to be breaking some colors on mobile:
https://github.com/wordpress-mobile/gutenberg-mobile/pull/2787#issuecomment-725027459

I noticed that after merging this PR, the `$gray-100` color defined on https://github.com/WordPress/gutenberg/blob/b571063c45e782f14677eeff6b74c4858882f3b9/packages/base-styles/_colors.scss#L13 takes preference and overwrites the one defined on https://github.com/WordPress/gutenberg/blob/b571063c45e782f14677eeff6b74c4858882f3b9/packages/base-styles/_colors.native.scss#L73 When previous this merge commit, the opposite was true.

We would expect that `*.native.scss` would always overwrite their non `native` counterparts.
Any direction on how to go around this issue?
cc @hypest @ceyhun

**Comment by etoledom** [comment]
We have found a plausible reason and a workaround to this issue here: https://github.com/WordPress/gutenberg/pull/26898

But still we would like to to find a proper solution soon. `_color.scss` should not be loaded if we have a `_color.native.scss` defined.

Any idea or direction is welcomed 🙏  
cc @yscik @gziolo

**Comment by yscik** [comment]
Looks like it's caused by this ~issue~ intentional feature: https://github.com/sass/dart-sass/issues/574: import statements with a relative path are not passed to the `importer` function for resolution. (Also see the compatibility note in [Sass docs](https://sass-lang.com/documentation/js-api#importer))

That pretty much breaks the whole concept of being able to specify a `.native.scss` override. Is it a common pattern to use the overrides this way? I see most of them are imported from the `.native.js` files directly, which should still work.

It could be pretty tricky to work around this limitation. Might be possible to add another preprocessing step for resolving the URLs in import statements before passing the content over to the Sass renderer?

For the specific case of the `colors.sccs` override it sounds like changing the import in `_variables.scss` from `@import './colors'` to `@import 'colors'` could also work.

**Pull request merged**
